### PR TITLE
bpo-46564: Optimize ``super().meth()`` calls via adaptive superinstructions

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
They should now have almost no overhead over a corresponding ``self.meth()`` call.

Summary of changes:
- ``typeobject.c`` -- refactoring to reuse code during specialization, also use ``InterpreterFrame`` over ``PyFrameObject`` for lazy frame benefits. Some changes here are partially taken from https://github.com/python/cpython/pull/24936. All credits to @vladima (I've tried to properly include them in the news item too.)
- ``specialize.c`` -- specialize for the 0-argument and 2-argument form of ``super()``.
- ``ceval.c`` -- does both a ``CALL`` and ``LOAD_METHOD`` without intermediates (and both are specialized forms too).

TODO:
benchmarks!

https://bugs.python.org/issue46564

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
Marking as draft as I need make this work with the new `CALL` convention.

File: Python/ceval.c [hunk 1]
```diff
@@ -5069,6 +5070,86 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
         }
 
+        TARGET(CALL_NO_KW_SUPER_0__LOAD_METHOD_CACHED) {
+            /* super().meth */
+            assert(_Py_OPCODE(next_instr[0]) == LOAD_METHOD_ADAPTIVE);
+            assert(_Py_OPCODE(next_instr[-2]) != PRECALL_METHOD);
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            _PyObjectCache *cache1 = &caches[-1].obj;
+            _PyAdaptiveEntry *lm_adaptive = &caches[-2].adaptive;
```
Comment (markshannon) [inline Python/ceval.c:5085]: Is this the cache entry for the following `LOAD_METHOD`?
If it is, can you add an assert that this is the correct cache entry for the `LOAD_METHOD`?

File: Python/ceval.c [hunk 2]
```diff
@@ -5069,6 +5070,86 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
+
+            /* CALL_NO_KW_SUPER */
+            PyObject *su_obj;
+            PyTypeObject *su_type;
+            PyObject *meth;
+            PyObject *super_callable = TOP();
+
+            DEOPT_IF(_PyType_CAST(super_callable) != &PySuper_Type, CALL);
+            /* super() - zero argument form */
+            if (_PySuper_GetTypeArgs(frame, frame->f_code, &su_type, &su_obj) < 0) {
```
Comment (markshannon) [inline Python/ceval.c:5097]: Can't we do this at specialization time? The number of locals, the index of "self", and whether it is a cell are all known then. Likewise the nature of `__class__` is also known.

File: Python/ceval.c [hunk 3]
```diff
@@ -5069,6 +5070,86 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
+                PyErr_Clear();
+                DEOPT_IF(1, CALL);
+            }
+            assert(su_obj != NULL);
+            DEOPT_IF(lm_adaptive->version != Py_TYPE(su_obj)->tp_version_tag, CALL);
+            DEOPT_IF(cache0->version != su_type->tp_version_tag, CALL);
```
Comment (markshannon) [inline Python/ceval.c:5103]: When can this fail? 
Isn't the next item in the MRO determined solely by `type(self)` and `__class__`, both of which are known at this point?

**Comment by markshannon** [comment]
Maybe we should merge https://github.com/python/cpython/pull/31002 first, as that PR is simpler.
It would also allow us to compare the performance of just the specialization, without the removal of frame object allocation.

**Comment by Fidget-Spinner** [comment]
> Maybe we should merge #31002 first, as that PR is simpler. It would also allow us to compare the performance of just the specialization, without the removal of frame object allocation.

👍

File: Python/ceval.c (hunk 3 above)
Comment (Fidget-Spinner) [reply Python/ceval.c:5103]: I wanted assurance that ``__class__`` didn't change.. Then again, I'm not sure if it can?

File: Python/ceval.c (hunk 1 above)
Comment (Fidget-Spinner) [reply Python/ceval.c:5085]: Yes it is. I'll add an assert.

I'm aware it's a little risky to use the next cache entry. If the hypothetical situation where CALL_SUPER_LOAD_METHOD deopts, and the next LOAD_METHOD then specializes, the subsequent call to CALL_SUPER_LOAD_METHOD will at best de-opt again, at worst erroneously allow invalid conditions to execute.

To protect against this, we need to ensure LOAD_METHOD will never specialize when it sees a prior super instruction. That way its memory will remain unused (and usable by CALL_SUPER_LOAD_METHOD.

**Comment by Fidget-Spinner** [comment]
Mark, I'm going to run benchmarks on `deltablue` first since it uses 2-argument form super. I'll address your optimization ideas for the 0-arg form once those results come back. Fingers crossed.

File: Objects/typeobject.c [hunk 4]
```diff
@@ -9011,8 +9045,8 @@ super_descr_get(PyObject *self, PyObject *obj, PyObject *type)
     }
 }
 
-static int
-super_init_without_args(InterpreterFrame *cframe, PyCodeObject *co,
+int
+_PySuper_GetTypeArgs(InterpreterFrame *cframe, PyCodeObject *co,
                         PyTypeObject **type_p, PyObject **obj_p)
```
Comment (arhadthedev) [inline Objects/typeobject.c:9050]:
```suggestion
                     PyTypeObject **type_p, PyObject **obj_p)
```

The line was aligned with an opening parenthesis of a parameter list.

File: Include/internal/pycore_code.h [hunk 5]
```diff
@@ -272,7 +275,8 @@ int _Py_Specialize_LoadMethod(PyObject *owner, _Py_CODEUNIT *instr, PyObject *na
 int _Py_Specialize_BinarySubscr(PyObject *sub, PyObject *container, _Py_CODEUNIT *instr, SpecializedCacheEntry *cache);
 int _Py_Specialize_StoreSubscr(PyObject *container, PyObject *sub, _Py_CODEUNIT *instr);
 int _Py_Specialize_CallNoKw(PyObject *callable, _Py_CODEUNIT *instr, int nargs,
-    PyObject *kwnames, SpecializedCacheEntry *cache, PyObject *builtins);
+     PyObject *kwnames, SpecializedCacheEntry *cache, PyObject *builtins,
+    PyObject **stack_pointer, InterpreterFrame *frame, PyObject *names);
```
Comment (arhadthedev) [inline Include/internal/pycore_code.h:279]:
```suggestion
    PyObject *kwnames, SpecializedCacheEntry *cache, PyObject *builtins,
    PyObject **stack_pointer, InterpreterFrame *frame, PyObject *names);
```

as in a removed line, or even:

```suggestion
                                              PyObject *kwnames, SpecializedCacheEntry *cache, PyObject *builtins,
                                              PyObject **stack_pointer, InterpreterFrame *frame, PyObject *names);
```

as in `_Py_Specialize_BinaryOp` right below.

**Review by arhadthedev — CHANGES_REQUESTED** [review]
A couple of indentation-related inconsistencies:

**Comment by Fidget-Spinner** [comment]
> Mark, I'm going to run benchmarks on `deltablue` first since it uses 2-argument form super. I'll address your optimization ideas for the 0-arg form once those results come back. Fingers crossed.

Well that was depressing. `deltablue` only shows 1.03x speedup. Looking closer at the code, `super` isn't called in any tight loops so that might be why. Maybe I need to pull out microbenchmarks now.

**Pull request closed without merging**